cli/context/store/store.go

```diff
@@ -259,12 +271,113 @@ func Export(name string, s Reader) io.ReadCloser {
 	return reader
 }
 
+const (
+	maxAllowedFileSizeToImport int64 = 10 << 20
+)
+
+type fileData interface {
+	get() ([]byte, error)
+}
+
+type zipFileData struct {
+	zf *zip.File
+}
+
+func (zfd *zipFileData) get() ([]byte, error) {
+	src, err := zfd.zf.Open()
+	if err != nil {
+		return nil, err
+	}
+	defer src.Close()
+	data, err := LimitedReadAll(src, maxAllowedFileSizeToImport)
+	if err != nil {
+		return nil, err
+	}
+	return data, nil
+}
+
+type tarFileData struct {
+	tr io.Reader
+}
+
+func (tfd *tarFileData) get() ([]byte, error) {
+	data, err := LimitedReadAll(tfd.tr, maxAllowedFileSizeToImport)
+	if err != nil {
+		return nil, err
+	}
+	return data, nil
+}
+
+func newZipFileData(zf *zip.File) fileData {
+	var zfd = new(zipFileData)
+	zfd.zf = zf
+	return zfd
+}
+
+func newTarFileData(tr io.Reader) fileData {
+	var tfd = new(tarFileData)
+	tfd.tr = tr
+	return tfd
+}
+
+// ImportType enum to represent import type
+type ImportType int
+
+const (
+	// Cli will import Docker context from current env
+	Cli ImportType = iota
+	// Tar will import Docker context from a tar file
+	Tar
+	// Zip will import Docker context from a zip file
+	Zip
+)
+
 // Import imports an exported context into a store
-func Import(name string, s Writer, reader io.Reader) error {
+func Import(name string, s Writer, reader io.Reader, importType ImportType) error {
+	switch importType {
+	case Zip:
+		return importZip(name, s, reader)
+	default: // Tar, Cli
+		return importTar(name, s, reader)
+	}
+}
+
+func importZip(name string, s Writer, reader io.Reader) error {
+	body, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return err
+	}
+	zr, err := zip.NewReader(bytes.NewReader(body), int64(len(body)))
+	if err != nil {
+		return err
+	}
+
+	tlsData := ContextTLSData{
+		Endpoints: map[string]EndpointTLSData{},
+	}
+
+	for _, zf := range zr.File {
+		if zf.FileInfo().IsDir() {
+			// skip this entry, only taking files into account
+			continue
+		}
+
+		fd := newZipFileData(zf)
```

Comment:
Sorry, I've been very busy, so didn't find time to look closely and comment, but this stood out to me: here we're protecting against a big file _inside_ the ZIP (same as for `tar` below). While it's definitely that someone with ill intent crafts a Zip file that's small, but contains a file that expands to a very large file, I think the main (or _only_) thing we should be concerned about is reading the zip/tar _itself_.

So, instead we could use a limitReader for

```go
body, err := ioutil.ReadAll(reader)
```

(to protect against someone accidentally doing `docker context import ./game-of-thrones-season-1-9.zip`)

And here just read the file as usual. Doing so would probably also reduce the complexity (we wouldn't need the `newZipFileData` and `newTarFileData`